File: SensorReader.cpp
#include "SensorReader.h"
#include "Logger.h"

SensorReader::SensorReader() {
    for (uint8_t i = 0; i < NUM_SENSORS; ++i) {
        _sensorData[i].distance = 0;
        _sensorData[i].valid = false;
    }
}

bool SensorReader::begin() {
    Wire.begin();
    Wire.setClock(400000); // 400 kHz

    Logger::println("=== Inicializando Sensores ToF vía XSHUT ===");

    // PASO 1: Configurar pines como salida y APAGAR todos los sensores
    for (uint8_t i = 0; i < NUM_SENSORS; ++i) {
        pinMode(XSHUT_PINS[i], OUTPUT);
        digitalWrite(XSHUT_PINS[i], LOW);
    }
    delay(100); // Dar tiempo a que todos se reinicien

    // PASO 2: Encender y configurar uno por uno
    for (uint8_t i = 0; i < NUM_SENSORS; ++i) {
        // Activamos el sensor actual poniendo su XSHUT en HIGH
        digitalWrite(XSHUT_PINS[i], HIGH);
        delay(10); 

        Logger::print("Iniciando Sensor ");
        Logger::print(i);
        Logger::print(" en pin ");
        Logger::print(XSHUT_PINS[i]);
        Logger::print(" con ángulo ");
        Logger::print(SENSOR_ANGLES[i]);
        Logger::print("°...");

        // Todos los sensores VL53L0X despiertan por defecto en la dirección de fábrica 0x29
        if (!_sensors[i].begin(0x29)) {
            Logger::println(" ¡ERROR! No responde en 0x29");
            return false;
        }

        // Cambiamos su dirección inmediatamente a su dirección única definitiva
        _sensors[i].setAddress(SENSOR_ADDRESSES[i]);
        
        Logger::print(" Asignada dirección en hexadecimal: ");
        // Imprimimos la dirección usando Serial directo para evitar el conflicto con el Logger
        #if ENABLE_SERIAL
        Serial.println(SENSOR_ADDRESSES[i], HEX);
        #endif
    }

    Logger::println("=== Todos los sensores inicializados con éxito ===");
    return true;
}

void SensorReader::readAll() {
    VL53L0X_RangingMeasurementData_t measurement;

    for (uint8_t i = 0; i < NUM_SENSORS; ++i) {
        // Ya no llamamos a _selectChannel. Leemos directamente del objeto.
        _sensors[i].rangingTest(&measurement, false);

        // Comprobamos si la lectura es válida (Status 4 suele ser error de fase/rango en Adafruit)
        if (measurement.RangeStatus != 4) {
            _sensorData[i].distance = measurement.RangeMilliMeter;
            
            if (_sensorData[i].distance > RELIABLE_RANGE) {
                _sensorData[i].distance = RELIABLE_RANGE;
            }
            
            _sensorData[i].valid = (_sensorData[i].distance >= MIN_VALID_DISTANCE);
        } else {
            _sensorData[i].distance = 0;
            _sensorData[i].valid = false;
        }
    }
}

const SensorData* SensorReader::getAllData() const { return _sensorData; }

File: src/clases/SensorReader.py
# sensor_reader.py
import time
from machine import Pin, I2C
from vl53l0x import VL53L0X
import Config
from Kalman import KalmanFilter1D  # Filtro Kalman 1D

class SensorData:

    def __init__(self):
        self.distance = 0
        self.valid = False

class SensorReader:
    def __init__(self):
        # Creamos el array de datos estáticos (como el _sensorData[NUM_SENSORS])
        self._sensor_data = [SensorData() for _ in range(Config.NUM_SENSORS)]
        self._sensors = []
        self._i2c = None
        self.historial_sensores = [[] for _ in range(Config.NUM_SENSORS)]
        self.kalman = KalmanFilter1D()
        self.VENTANA_FILTRO = 4 #Hacer variable de Config o variable de clase.

    def begin(self):
        # Inicializa I2C a 400 kHz (Wire.setClock(400000))
        # Ajusta pin SCL y SDA según tus conexiones en el ESP32-S2 Mini
        self._i2c = I2C(0, scl=Pin(9), sda=Pin(8), freq=100000)

        print("=== Inicializando Sensores ToF vía XSHUT ===")

        # PASO 1: Configurar pines como salida y APAGAR todos los sensores (LOW)
        xshut_objects = []
        for i in range(Config.NUM_SENSORS):
            pin_out = Pin(Config.XSHUT_PINS[i], Pin.OUT)
            pin_out.value(0)  # digitalWrite(XSHUT_PINS[i], LOW)
            xshut_objects.append(pin_out)
        
        time.sleep_ms(100)  # delay(100) para el reset de hardware

        # PASO 2: Encender y configurar uno por uno
        for i in range(Config.NUM_SENSORS):
            # Activamos el sensor actual poniendo su XSHUT en HIGH (1)
            xshut_objects[i].value(1)
            time.sleep_ms(10)  # delay(10)

            print(f"Iniciando Sensor {i} en pin {Config.XSHUT_PINS[i]} con ángulo {Config.SENSOR_ANGLES[i]}°...")

            try:
                # Todos los sensores despiertan en 0x29 de fábrica
                sensor_instancia = VL53L0X(self._i2c, address=0x29)
                
                # Cambiamos su dirección inmediatamente a su dirección única definitiva
                sensor_instancia.set_address(Config.SENSOR_ADDRESSES[i])
                
                self._sensors.append(sensor_instancia)
                
                if Config.ENABLE_SERIAL:
                    print(f" Asignada dirección en hexadecimal: {hex(Config.SENSOR_ADDRESSES[i])}")
                    
            except Exception as e:
                print(" ¡ERROR! No responde en 0x29")
                return False

        print("=== Todos los sensores inicializados con éxito ===")
        return True

    def read_all(self):
        for i in range(Config.NUM_SENSORS):
            try:
                # Llama a la lectura del sensor
                distance_mm = self._sensors[i].read()

                # Comprobamos si la lectura es válida
                # El driver devuelve 0 si hay error de fase o de lectura (Equivalente al RangeStatus != 4)
                if distance_mm > 0:
                    
                    # Filtro de seguridad cap (RELIABLE_RANGE)
                    if distance_mm > Config.RELIABLE_RANGE:
                        distance_mm = Config.RELIABLE_RANGE
                    
                    self._sensor_data[i].distance = distance_mm
                    self._sensor_data[i].valid = (distance_mm >= Config.MIN_VALID_DISTANCE)
                else:
                    # Lectura inválida (Status de error)
                    self._sensor_data[i].distance = 0
                    self._sensor_data[i].valid = False
                    
            except Exception:
                # Protección si falla la línea física I2C en pleno bucle
                self._sensor_data[i].distance = 0
                self._sensor_data[i].valid = False

    def get_all_data(self):
        # Equivalente al const SensorData* getAllData()
        return self._sensor_data
    
    def get_filtered_data(self):
        """
        Recorre sensor a sensor usando un ciclo for interno.
        Devuelve una lista limpia de 5 elementos con distancias promediadas.
        """
        distancias_filtradas = []
        for i in range(Config.NUM_SENSORS):
            lectura_cruda = self._sensor_data[i].distance
            
            # Descarte de picos erróneos (Outliers)
            if lectura_cruda > 4000 or lectura_cruda < 0:
                if len(self.historial_sensores[i]) > 0:
                    lectura_cruda = self.historial_sensores[i][-1]
                else:
                    lectura_cruda = 0
            
            self.historial_sensores[i].append(lectura_cruda)
            if len(self.historial_sensores[i]) > self.VENTANA_FILTRO:
                self.historial_sensores[i].pop(0)
                
            promedio = sum(self.historial_sensores[i]) / len(self.historial_sensores[i])
            distancias_filtradas.append(promedio)
            
        return distancias_filtradas
